File: clients/python/opensysml/diagnostic.py
"""Diagnostic class wrapping protobuf diagnostic message."""


class Diagnostic:
    """Represents a parse diagnostic (error or warning).
    
    Attributes:
        severity (str): Diagnostic severity (e.g., "error", "warning")
        message (str): Diagnostic message
        code (str): Stable identifier of what was found (a validation code,
            "syntax", "choice-point", "guard-unevaluable"); "" when none
        file (str): Source file name
        start_line (int): Starting line number (1-based)
        start_column (int): Starting column number (1-based)
        end_line (int): Ending line number (1-based)
        end_column (int): Ending column number (1-based)
        span: Protobuf Span object
    """
    
    def __init__(self, pb_diagnostic):
        """Initialize Diagnostic from protobuf Diagnostic message.
        
        Args:
            pb_diagnostic: sysml_pb2.Diagnostic protobuf message
        """
        self._pb = pb_diagnostic
    
    @property
    def severity(self):
        """Get diagnostic severity."""
        return self._pb.severity
    
    @property
    def message(self):
        """Get diagnostic message."""
        return self._pb.message
    
    @property
    def code(self):
        """Get the diagnostic code, independent of the message wording."""
        return self._pb.code
    
    @property
    def span(self):
        """Get protobuf Span object."""
        return self._pb.span
    
    @property
    def file(self):
        """Get source file name."""
        return self._pb.span.file
    
    @property
    def start_line(self):
        """Get starting line number (1-based)."""
        return self._pb.span.start_line
    
    @property
    def start_column(self):
        """Get starting column number (1-based)."""
        return self._pb.span.start_col
    
    @property
    def end_line(self):
        """Get ending line number (1-based)."""
        return self._pb.span.end_line
    
    @property
    def end_column(self):
        """Get ending column number (1-based)."""
        return self._pb.span.end_col
    
    def __str__(self):
        """String representation: 'file:line:col: severity: message'."""
        return (
            f"{self.file}:{self.start_line}:{self.start_column}: "
            f"{self.severity}: {self.message}"
        )
    
    def __repr__(self):
        """Detailed representation."""
        return (
            f"Diagnostic(severity={self.severity!r}, "
            f"message={self.message!r}, "
            f"code={self.code!r}, "
            f"file={self.file!r}, "
            f"line={self.start_line})"
        )

File: clients/java/opensysml-client/src/main/java/org/openmbee/opensysml/Diagnostic.java
package org.openmbee.opensysml;

import java.util.Locale;
import java.util.Objects;
import java.util.Optional;

/**
 * A parse or semantic finding the service reported about a model.
 *
 * @param severity how serious the finding is
 * @param message the finding, as the service words it
 * @param code stable identifier of what was found, independent of the message wording: a
 *     validation code, {@code "syntax"}, or {@code "choice-point"} / {@code "guard-unevaluable"}
 *     on a run; empty when the service assigned none
 * @param span where in the source it is, absent when the service located none
 */
public record Diagnostic(Severity severity, String message, String code, Optional<Span> span) {

  /**
   * Creates a diagnostic.
   *
   * @param severity the severity, never {@code null}
   * @param message the message, never {@code null}
   * @param code the code, empty rather than {@code null} when there is none
   * @param span the source location, absent when unlocated
   */
  public Diagnostic {
    Objects.requireNonNull(severity, "severity");
    Objects.requireNonNull(message, "message");
    Objects.requireNonNull(code, "code");
    Objects.requireNonNull(span, "span");
  }

  /** How serious a diagnostic is. */
  public enum Severity {
    /** The model is wrong: parsing or validation failed. */
    ERROR,
    /** The model is questionable but was accepted. */
    WARNING,
    /** Informational only. */
    INFO,
    /** A severity this release of the client does not know. */
    UNKNOWN;

    /**
     * The severity a service reported, by its wire name.
     *
     * @param wireName the severity as the service spells it ({@code "error"})
     * @return the matching severity, or {@link #UNKNOWN} for anything else
     */
    public static Severity fromWireName(String wireName) {
      return switch (wireName) {
        case "error" -> ERROR;
        case "warning" -> WARNING;
        case "info" -> INFO;
        default -> UNKNOWN;
      };
    }

    /**
     * The name the service spells this severity with.
     *
     * @return the wire name; {@code "unknown"} for a severity this release does not know
     */
    public String wireName() {
      return name().toLowerCase(Locale.ROOT);
    }
  }

  /**
   * A range of a source file.
   *
   * @param file path of the file, as the service reports it
   * @param startLine first line, 1-based
   * @param startColumn first column, 1-based
   * @param endLine last line, 1-based
   * @param endColumn last column, 1-based
   */
  public record Span(String file, int startLine, int startColumn, int endLine, int endColumn) {
    /**
     * Creates a span.
     *
     * @param file the file, never {@code null}
     * @param startLine first line
     * @param startColumn first column
     * @param endLine last line
     * @param endColumn last column
     */
    public Span {
      Objects.requireNonNull(file, "file");
    }
  }
}
